opm/simulators/wells/BlackoilWellModelGuideRates.cpp

```diff
@@ -0,0 +1,588 @@
+  Copyright 2016 - 2018 IRIS AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/simulators/wells/BlackoilWellModelGuideRates.hpp>
+
+#include <opm/input/eclipse/Schedule/Schedule.hpp>
+#include <opm/input/eclipse/Schedule/Group/Group.hpp>
+
+#include <opm/output/data/Groups.hpp>
+#include <opm/output/data/GuideRateValue.hpp>
+
+#include <opm/simulators/wells/BlackoilWellModelGeneric.hpp>
+#include <opm/simulators/wells/WellGroupHelpers.hpp>
+#include <opm/simulators/wells/WellInterfaceGeneric.hpp>
+
+#include <fmt/format.h>
+
+namespace {
+
+struct RetrieveWellGuideRate
+{
+    RetrieveWellGuideRate() = default;
+
+    explicit RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                   const std::string&    wgname);
+
+    explicit RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                   const Opm::Group&     group);
+
+    bool prod      { false };
+    bool inj_water { false };
+    bool inj_gas   { false };
+};
+
+RetrieveWellGuideRate
+operator||(RetrieveWellGuideRate lhs, const RetrieveWellGuideRate& rhs)
+{
+    lhs.prod      = lhs.prod      || rhs.prod;
+    lhs.inj_water = lhs.inj_water || rhs.inj_water;
+    lhs.inj_gas   = lhs.inj_gas   || rhs.inj_gas;
+
+    return lhs;
+}
+
+RetrieveWellGuideRate::RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                             const std::string&    wgname)
+    : prod      { guideRate.has(wgname) }
+    , inj_water { guideRate.has(wgname, Opm::Phase::WATER) }
+    , inj_gas   { guideRate.has(wgname, Opm::Phase::GAS)   }
+{}
+
+RetrieveWellGuideRate::RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                             const Opm::Group&     group)
+    : RetrieveWellGuideRate{ guideRate, group.name() }
+{
+    if (group.isProductionGroup()) {
+        this->prod = true;
+    }
+
+    if (group.isInjectionGroup()) {
+        this->inj_water = this->inj_water || group.hasInjectionControl(Opm::Phase::WATER);
+        this->inj_gas   = this->inj_gas   || group.hasInjectionControl(Opm::Phase::GAS);
+    }
+}
+
+class GroupTreeWalker
+{
+public:
+    using GroupOp = std::function<void(const Opm::Group&)>;
+    using WellOp = std::function<void(const Opm::Well&)>;
```

Comment:
We must `#include <functional>` to use `std:function<>` and `std::reference_wrapper<>`.

Reply:
done